src/java/org/apache/cassandra/db/commitlog/CommitLogSegment.java

```diff
@@ -764,4 +763,18 @@ public CommitLogPosition getCommitLogPosition()
             return new CommitLogPosition(segment.id, buffer.limit());
         }
     }
+
+    protected abstract static class Builder<S extends CommitLogSegment>
+    {
+        protected final AbstractCommitLogSegmentManager segmentManager;
+
+        public Builder(AbstractCommitLogSegmentManager segmentManager)
+        {
+            this.segmentManager = segmentManager;
+        }
+
+        public abstract S build();
```

Comment:
The generification here is not helpful and only clutters the code or causes raw type warnings. Each subclass can return its concrete type here, and a caller can only benefit from the generic if it already knows the concrete type, in which case it is calling a method that does return the concrete type.

Reply:
indeed